File: pymoose/PyRun.cpp
// PyRun.cpp ---
//
// Filename: PyRun.cpp
// Description:
// Author: subha
// Maintainer:
// Created: Sat Oct 11 14:47:22 2014 (+0530)
// Version:
// Last-Updated: Fri Jun 19 18:56:06 2015 (-0400)
//           By: Subhasis Ray
//     Update #: 15
// URL:
// Keywords:
// Compatibility:
//
//

// Commentary:
//
//
//
//

// Change log:
//
//
//
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as
// published by the Free Software Foundation; either version 3, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; see the file COPYING.  If not, write to
// the Free Software Foundation, Inc., 51 Franklin Street, Fifth
// Floor, Boston, MA 02110-1301, USA.
//
//

// Code:

#include "Python.h"
#include "../basecode/header.h"
#include "PyRun.h"

const int PyRun::RUNPROC = 1;
const int PyRun::RUNTRIG = 2;
const int PyRun::RUNBOTH = 0;

static SrcFinfo1< double >* outputOut()
{
    static SrcFinfo1< double > outputOut( "output",
                                          "Sends out the value of local variable called `output`. Thus, you can"
                                          " have Python statements which compute some value and assign it to the"
                                          " variable called `output` (which is defined at `reinit` call). This"
                                          " will be sent out to any target connected to the `output` field.");
    return &outputOut;
}


const Cinfo * PyRun::initCinfo()
{
    static ValueFinfo< PyRun, string > runstring(
        "runString",
        "String to be executed at each time step.",
        &PyRun::setRunString,
        &PyRun::getRunString);

    static ValueFinfo< PyRun, string > initstring(
        "initString",
        "String to be executed at initialization (reinit).",
        &PyRun::setInitString,
        &PyRun::getInitString);

    static ValueFinfo< PyRun, string > inputvar(
        "inputVar",
        "Name of local variable in which input balue is to be stored. Default"
        " is `input_` (to avoid conflict with Python's builtin function"
        " `input`).",
        &PyRun::setInputVar,
        &PyRun::getInputVar);

    static ValueFinfo< PyRun, string > outputvar(
        "outputVar",
        "Name of local variable for storing output. Default is `output`.",
        &PyRun::setOutputVar,
        &PyRun::getOutputVar);

    static ValueFinfo< PyRun, int > mode(
        "mode",
        "Flag to indicate whether runString should be executed for both trigger and process, or one of them.",
        &PyRun::setMode,
        &PyRun::getMode);

    // static ValueFinfo< PyRun, PyObject* > globals(
    //     "globals",
    //     "Global environment dict",
    //     &PyRun::setGlobals,
    //     &PyRun::getGlobals);

    // static ValueFinfo< PyRun, PyObject* > locals(
    //     "locals",
    //     "Local environment dict",
    //     &PyRun::setLocals,
    //     &PyRun::getLocals);

    static DestFinfo trigger(
        "trigger",
        "Executes the current runString whenever a message arrives. It stores"
        " the incoming value in local variable named"
        " `input_`, which can be used in the"
        " `runString` (the underscore is added to avoid conflict with Python's"
        " builtin function `input`). If debug is True, it prints the input"
        " value.",
        new EpFunc1< PyRun, double >(&PyRun::trigger));

    static DestFinfo run(
        "run",
        "Runs a specified string. Does not modify existing run or init strings.",
        new EpFunc1< PyRun, string >(&PyRun::run));

    static DestFinfo process(
        "process",
        "Handles process call. Runs the current runString.",
        new ProcOpFunc< PyRun >(&PyRun::process));

    static DestFinfo reinit(
        "reinit",
        "Handles reinit call. Runs the current initString.",
        new ProcOpFunc< PyRun >( &PyRun::reinit ));

    static Finfo * processShared[] = { &process, &reinit };
    static SharedFinfo proc(
        "proc",
        "This is a shared message to receive Process messages "
        "from the scheduler objects."
        "The first entry in the shared msg is a MsgDest "
        "for the Process operation. It has a single argument, "
        "ProcInfo, which holds lots of information about current "
        "time, thread, dt and so on. The second entry is a MsgDest "
        "for the Reinit operation. It also uses ProcInfo. ",
        processShared, sizeof( processShared ) / sizeof( Finfo* ));

    static Finfo * pyRunFinfos[] = {
        &runstring,
        &initstring,
        &mode,
        &inputvar,
        &outputvar,
        &trigger,
        outputOut(),
        // &locals,
        // &globals,
        &run,
        &proc,
    };

    static string doc[] = {
        "Name", "PyRun",
        "Author", "Subhasis Ray",
        "Description", "Runs Python statements from inside MOOSE."};
    static Dinfo< PyRun > dinfo;
    static Cinfo pyRunCinfo(
        "PyRun",
        Neutral::initCinfo(),
        pyRunFinfos,
        sizeof(pyRunFinfos) / sizeof(Finfo*),
        &dinfo,
        doc,
        sizeof(doc) / sizeof(string));
    return &pyRunCinfo;
}

static const Cinfo * pyRunCinfo = PyRun::initCinfo();

PyRun::PyRun():mode_(0), initstr_(""), runstr_(""),
               globals_(0), locals_(0),
               runcompiled_(0), initcompiled_(0),
               inputvar_("input_"), outputvar_("output")
{
    locals_ = PyDict_New();
    if (!locals_){
        cerr << "Could not initialize locals dict" << endl;
        return;
    }
    PyObject * value = PyFloat_FromDouble(0.0);
    if (!value && PyErr_Occurred()){
        PyErr_Print();
        return;
    }
    if (PyDict_SetItemString(locals_, inputvar_.c_str(), value)){
        PyErr_Print();
    }
}

PyRun::~PyRun()
{
    Py_XDECREF(globals_);
    Py_XDECREF(locals_);
}

void PyRun::setRunString(string statement)
{
    runstr_ = statement;
}

string PyRun::getRunString() const
{
    return runstr_;
}

void PyRun::setInitString(string statement)
{
    initstr_ = statement;
}

string PyRun::getInitString() const
{
    return initstr_;
}

void PyRun::setInputVar(string name)
{
    PyDict_DelItemString(locals_, inputvar_.c_str());
    inputvar_ = name;
}

string PyRun::getInputVar() const
{
    return inputvar_;
}

void PyRun::setOutputVar(string name)
{
    PyDict_DelItemString(locals_, outputvar_.c_str());
    outputvar_ = name;
}

string PyRun::getOutputVar() const
{
    return outputvar_;
}

void PyRun::setMode(int flag)
{
    mode_ = flag;
}

int PyRun::getMode() const
{
    return mode_;
}

void PyRun::trigger(const Eref& e, double input)
{
    if (!runcompiled_){
        return;
    }
    if (mode_ == 1){
        return;
    }

    PyObject * value = PyDict_GetItemString(locals_, inputvar_.c_str());
    if (value){
        Py_DECREF(value);
    }
    value = PyFloat_FromDouble(input);
    if (!value && PyErr_Occurred()){
        PyErr_Print();
    }
    if (PyDict_SetItemString(locals_, inputvar_.c_str(), value)){
        PyErr_Print();
    }
    PyEval_EvalCode(runcompiled_, globals_, locals_);
    if (PyErr_Occurred()){
        PyErr_Print ();
    }
    value = PyDict_GetItemString(locals_, outputvar_.c_str());
    if (value){
        double output = PyFloat_AsDouble(value);
        if (PyErr_Occurred()){
            PyErr_Print ();
        } else {
            outputOut()->send(e, output);
        }
    }
}

void PyRun::run(const Eref&e, string statement)
{
    PyRun_SimpleString(statement.c_str());
    PyObject * value = PyDict_GetItemString(locals_, outputvar_.c_str());
    if (value){
        double output = PyFloat_AsDouble(value);
        if (PyErr_Occurred()){
            PyErr_Print ();
        } else {
            outputOut()->send(e, output);
        }
    }
}

void PyRun::process(const Eref & e, ProcPtr p)
{
    // PyRun_String(runstr_.c_str(), 0, globals_, locals_);
    // PyRun_SimpleString(runstr_.c_str());
    if (!runcompiled_ || mode_ == 2){
        return;
    }
    PyEval_EvalCode(runcompiled_, globals_, locals_);
    if (PyErr_Occurred()){
        PyErr_Print ();
    }
    PyObject * value = PyDict_GetItemString(locals_, outputvar_.c_str());
    if (value){
        double output = PyFloat_AsDouble(value);
        if (PyErr_Occurred()){
            PyErr_Print ();
        } else {
            outputOut()->send(e, output);
        }
    }
}

/**
   This is derived from:
   http://effbot.org/pyfaq/how-do-i-tell-incomplete-input-from-invalid-input.htm
 */
void handleError(bool syntax)
{
    PyObject *exc, *val, *trb;
    char * msg;

    if (syntax && PyErr_ExceptionMatches (PyExc_SyntaxError)){
        PyErr_Fetch (&exc, &val, &trb);        /* clears exception! */

        if (PyArg_ParseTuple (val, "sO", &msg, &trb) &&
            !strcmp (msg, "unexpected EOF while parsing")){ /* E_EOF */
            Py_XDECREF (exc);
            Py_XDECREF (val);
            Py_XDECREF (trb);
        } else {                                  /* some other syntax error */
            PyErr_Restore (exc, val, trb);
            PyErr_Print ();
        }
    } else {                                     /* some non-syntax error */
        PyErr_Print ();
    }
}

void PyRun::reinit(const Eref& e, ProcPtr p)
{
    PyObject * main_module;
    if (globals_ == NULL){
        main_module = PyImport_AddModule("__main__");
        globals_ = PyModule_GetDict(main_module);
        Py_XINCREF(globals_);
    }
    if (locals_ == NULL){
        locals_ = PyDict_New();
        if (!locals_){
            cerr << "Could not initialize locals dict" << endl;
        }
    }
    initcompiled_ = (PYCODEOBJECT*)Py_CompileString(
        initstr_.c_str(),
        get_program_name().c_str(),
        Py_file_input);
    if (!initcompiled_){
        cerr << "Error compiling initString" << endl;
        handleError(true);
    } else {
        PyEval_EvalCode(initcompiled_, globals_, locals_);
        if (PyErr_Occurred()){
            PyErr_Print ();
        }
    }
    runcompiled_ = (PYCODEOBJECT*)Py_CompileString(
        runstr_.c_str(),
        get_program_name().c_str(),
        Py_file_input);
    if (!runcompiled_){
        cerr << "Error compiling runString" << endl;
        handleError(true);
    } else {
        PyEval_EvalCode(runcompiled_, globals_, locals_);
        if (PyErr_Occurred()){
            PyErr_Print ();
        }
    }
}


//
// PyRun.cpp ends here

File: moose-examples/snippets/pyrun.py
# pyrun.py ---
#
# Filename: pyrun.py
# Description:
# Author: Subhasis Ray
# Maintainer:
# Created: Wed Oct 15 10:14:15 2014 (+0530)
# Version:
# Last-Updated:
#           By:
#     Update #: 0
# URL:
# Keywords:
# Compatibility:
#
#

# Commentary:
#
#
#
#

# Change log:
#
#
#
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 3, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street, Fifth
# Floor, Boston, MA 02110-1301, USA.
#
#

import numpy as np
from matplotlib import pyplot as plt
import moose

def run_sequence():
    """
    In this example we demonstrate the use of PyRun objects to execute
    Python statements from MOOSE. Here is a couple of fun things to
    indicate the power of MOOSE-Python integration.

    First we create a PyRun object called `Hello`. In its `initString`
    we put in Python statements that prints the element's string
    representation using pymoose-API. When ``moose.reinit()`` is
    called, this causes MOOSE to execute these Python statements which
    include Python calling a MOOSE function
    (Python->MOOSE->Python->MOOSE) - isn't that cool!

    We also initialize a counter called `hello_count` to 0.

    The statements in initString gets executed once, when we call
    ``moose.reinit()``.

    In the `runString` we put a couple of print statements to indicate
    the name of the object which is running and the current
    count. Then we increase the count directly.

    When we call ``moose.start()``, the `runString` gets executed at
    each time step.

    The other PyRun object we create, is `/World`. In its `initString`
    apart from ordinary print statements and initialization, we define
    a Python function called ``incr_count``. This silly little
    function just increments the global `world_count` by 1.

    The `runString` for `World` simply calls this function to
    increment the count and print it.

    We may notice that we assign tick 0 to `Hello` and tick 1 to
    `World`. Looking at the output, you will realize that the
    sequences of the ticks strictly maintain the sequence of
    execution.

    """
    model = moose.Neutral('/model')
    hello_runner = moose.PyRun('/model/Hello')
    hello_runner.initString = """
print( 'Init', moose.element('/model/Hello') )
hello_count = 0
"""
    hello_runner.runString = """
print( 'Running Hello' )
print( 'Hello count =', hello_count )
hello_count += 1
"""
    hello_runner.run('from datetime import datetime')
    hello_runner.run('print("Hello: current time:", datetime.now().isoformat())')
    moose.useClock(0, hello_runner.path, 'process')
    world_runner = moose.PyRun('World')
    world_runner.initString = """
print( 'Init World' )
world_count = 0
def incr_count():
    global world_count
    world_count += 1
"""
    world_runner.runString = """
print( 'Running World' )
print( 'World count =', world_count )
incr_count()
"""
    world_runner.run('from datetime import datetime')
    world_runner.run('print( "World: current time:", datetime.now().isoformat())')

    moose.useClock(0, world_runner.path, 'process')
    moose.reinit()
    moose.start(0.001)


def input_output():
    """
    The PyRun class can take a double input through `trigger`
    field. Whenever another object sends an input to this field, the
    `runString` is executed.

    The fun part of this is that you can use the input value in your
    python statements in `runString`. This is stored in a local
    variable called `input_`. You can rename this by setting `inputVar`
    field.

    Things become even more interesting when you can send out a value
    computed using Python. PyRun objects allow you to define a local
    variable called `output` and whatever value you assign to this,
    will be sent out through the source field `output` on successful
    execution of the `runString`.

    You can rename the output variable by setting `outputVar` field.

    In this example, we send the output of a pulsegen object sending
    out the values 1, 2, 3 during each pulse and compute the square of
    these numbers in Python and set output to this square.

    The calculated value is assigned to the `output` variable and in
    turn sent out to a Table object's input and gets recorded.

    By default PyRun executes the `runString` whenever a `trigger`
    message is received and when its process method is called at each
    timestep. In both cases it sends out the `output` value. Since
    this may cause inaccuracies depending on what the Python
    statements in `runString` do, a `mode` can be specified to disable
    one of the above. We set ``mode = 2`` to disable the `process`
    method. Note that this could also have been done by setting its
    ``tick = -1``.

    ``mode = 1`` will disable `trigger` message and ``mode = 0``, the
    default, enables both.
    """
    model = moose.Neutral('/model')
    input_pulse = moose.PulseGen('/model/pulse')
    #: set the baseline output 0
    input_pulse.baseLevel = 0.0
    #: We make it generate three pulses
    input_pulse.count = 3
    input_pulse.level[0] = 1.0
    input_pulse.level[1] = 2.0
    input_pulse.level[2] = 3.0
    #: Each pulse will appear 1 s after the previous one
    input_pulse.delay[0] = 1.0
    input_pulse.delay[1] = 1.0
    input_pulse.delay[2] = 1.0
    #: Each pulse is 1 s wide
    input_pulse.width[0] = 1.0
    input_pulse.width[1] = 1.0
    input_pulse.width[2] = 1.0
    #: Now create the PyRun object
    pyrun = moose.PyRun('/model/pyrun')
    pyrun.runString = """
output = input_ * input_
print( 'input =', input_ )
print( 'output =', output )
"""
    pyrun.mode = 2 # do not run process method
    moose.connect(input_pulse, 'output', pyrun, 'trigger')
    output_table = moose.Table('/model/output')
    moose.connect(pyrun, 'output', output_table, 'input')
    input_table = moose.Table('/model/input')
    moose.connect(input_pulse, 'output', input_table, 'input')
    moose.setClock(0, 0.25)
    moose.setClock(1, 0.25)
    moose.setClock(2, 0.25)
    moose.useClock(0, input_pulse.path, 'process')
    #: this is unnecessary because the mode=2 ensures that `process`
    #: does nothing
    moose.useClock(1, pyrun.path, 'process')
    moose.useClock(2, '/model/#[ISA=Table]', 'process')
    moose.reinit()
    moose.start(10.0)
    #ts =
    plt.plot(input_table.vector, label='input')
    plt.plot(output_table.vector, label='output')
    plt.legend()
    plt.show()

def main():
	"""
You can use the PyRun class to run Python statements from MOOSE at
runtime. This opens up many possibilities of interleaving computing in
Python and MOOSE. You can also use this for debugging simulations.
	"""
	run_sequence()
	moose.delete('/model')
	input_output()

if __name__ == '__main__':
    run_sequence()
    moose.delete('/model')
    input_output()



#
# pyrun.py ends here
